driver-core/src/test/unit/com/mongodb/internal/connection/DefaultConnectionPoolSpecification.groovy

```diff
@@ -287,6 +289,54 @@ class DefaultConnectionPoolSpecification extends Specification {
         1 * listener.connectionCheckedIn { it.connectionId.serverId == SERVER_ID }
     }
 
+    def 'should fire connection checkout failed with Reason.CONNECTION_ERROR if fails to open a connection'() {
+        given:
+        def listener = Mock(ConnectionPoolListener)
+        def connection = Mock(InternalConnection)
+        connection.getDescription() >> new ConnectionDescription(SERVER_ID)
+        connection.opened() >> false
+        connection.open() >> { throw new UncheckedIOException('expected failure', new IOException()) }
+        connectionFactory.create(SERVER_ID) >> connection
+        pool = new DefaultConnectionPool(SERVER_ID, connectionFactory, builder().addConnectionPoolListener(listener).build())
+
+        when:
+        try {
+            pool.get()
+        } catch (UncheckedIOException e) {
+            if ('expected failure' != e.getMessage()) {
+                throw e;
+            }
+        }
+
+        then:
+        1 * listener.connectionCheckOutFailed { it.reason == ConnectionCheckOutFailedEvent.Reason.CONNECTION_ERROR }
+    }
+
+    def 'should fire connection checkout failed with Reason.CONNECTION_ERROR if fails to open a connection asynchronously'() {
+        given:
+        def listener = Mock(ConnectionPoolListener)
+        def connection = Mock(InternalConnection)
+        connection.getDescription() >> new ConnectionDescription(SERVER_ID)
+        connection.opened() >> false
+        connection.openAsync(_) >> { SingleResultCallback<Void> callback ->
```

Comment:
Stubbing here is done with accordance to https://spockframework.org/spock/docs/1.3/all_in_one.html#_computing_return_values

Reply:
See: https://spockframework.org/spock/docs/2.0-M3/

Reply:
Ross, how did you find this link? Both the "Documentation" link on https://spockframework.org and the "Reference Documentation" link on https://github.com/spockframework/spock redirect to https://spockframework.org/spock/docs/1.3/index.html.

Reply:
I found it here: https://github.com/spockframework/spock/releases

Reply:
Thanks 👍